### PR TITLE
[AutoScheduler] Support layout rewrite for whole networks

### PR DESCRIPTION
**Pull request by merrymercy**
Auto-scheduler can infer a good layout for constant tensors (e.g., the weight tensors in conv2d, dense) according to the loop structure created by the schedule. This PR adds a relay pass to support this kind of layout rewrite at graph level. This pass inserts necessary layout transforms in relay program and pre-computes them by `FoldConstant` pass.

Changes overview
- Add a relay op `auto_scheduler_layout_transform` and its topi compute definition. This op can handle more general layout transform than the existing `relay.op.transform.layout_transform`
- Add a relay pass `AutoSchedulerLayoutRewrite` to do the layout rewrite.
- Add a field `auto_scheduler_rewritten_layout` to `Conv2dAttrs` to indicate the new layout after rewriting.
- see `python/tvm/topi/nn/conv2d.py` for examples of how to enable the layout rewrite for a TOPI compute.

co-authored by @minminsun

### CODE REVIEW

**Comment by merrymercy** [comment]
cc @tqchen @jcf94 @minminsun @FrozenGene @comaniac

**Comment by merrymercy** [comment]
@jcf94 @comaniac @FrozenGene @junrushao1994  Let us merge this as soon as possible, so we can get good performance on CPU and publish the CPU tutorial.
I will do more clean up in the follow-up PRs when I add layout rewrite support for more ops (conv3d, dense, batch matmul).
I will figure out the best way to modify TOPI compute when I add the support for more ops.

File: python/tvm/topi/nn/conv2d.py [hunk 1]
```diff
@@ -371,8 +379,29 @@ def conv2d_nhwc(Input, Filter, stride, padding, dilation, out_dtype="float32"):
     else:
         dilation_h, dilation_w = dilation
 
+    if auto_scheduler_rewritten_layout:
+        # Infer shape for the rewritten layout
```
Comment (FrozenGene) [inline python/tvm/topi/nn/conv2d.py:383]: As discussed, we need extract it. If conveniently, we could mark this one todo.

File: tests/python/relay/test_auto_scheduler_layout_rewrite.py [hunk 2]
```diff
@@ -0,0 +1,122 @@
...
+
+
+def tune_and_check(mod, data, weight):
+    # Extract tasks from a relay program
+    target = tvm.target.Target("llvm")
+    tasks, task_weights = auto_scheduler.extract_tasks(mod, target=target, params={})
+
+    with tempfile.NamedTemporaryFile() as fp:
+        log_file = fp.name
+        # log_file = "test_layout_rewrite.json"
```
Comment (FrozenGene) [inline tests/python/relay/test_auto_scheduler_layout_rewrite.py:80]: remove it. If we want to merge it soon, could leave it in the clean up pr.

**Review by FrozenGene — APPROVED** [review]

File: include/tvm/ir/transform.h [hunk 3]
```diff
@@ -197,6 +197,13 @@ class PassContext : public ObjectRef {
    */
   TVM_DLL void Trace(const IRModule& module, const PassInfo& info, bool is_before) const;
 
+  /*!
+   * \brief Check if a pass is enabled.
```
Comment (merrymercy) [inline include/tvm/ir/transform.h:201]:
```suggestion
   * \brief Check whether a pass is enabled.
```

File: tests/python/relay/test_auto_scheduler_layout_rewrite.py (hunk 2 above)
Comment (merrymercy) [inline tests/python/relay/test_auto_scheduler_layout_rewrite.py:80]:
```suggestion
```

File: python/tvm/topi/nn/conv2d.py (hunk 1 above)
Comment (merrymercy) [inline python/tvm/topi/nn/conv2d.py:383]:
```suggestion
        # Infer shape for the rewritten layout
        # todo(merrymercy): wrap this with a more general interface.
```

File: include/tvm/relay/transform.h [hunk 4]
```diff
@@ -106,6 +106,14 @@ TVM_DLL Pass FoldConstant();
  */
 TVM_DLL Pass FuseOps(int fuse_opt_level = -1);
 
+/*!
+ * \brief The inverse operation of FuseOps. It transforms a fused program returned by
+ * FuseOps into the program before FuseOps. (i.e., x == DefuseOps(FuseOps(x)))
```
Comment (merrymercy) [inline include/tvm/relay/transform.h:111]:
```suggestion
 * FuseOps into the program before FuseOps. (i.e. x == DefuseOps(FuseOps(x)))
```

File: tests/python/relay/test_auto_scheduler_layout_rewrite.py [hunk 5]
```diff
@@ -0,0 +1,121 @@
...
+        data_layout=layout,
+        kernel_layout=kernel_layout,
+    )
+    mod = tvm.IRModule()
+    mod["main"] = relay.Function([d, w], y)
+    data, weight = get_np_array(d, dtype), get_np_array(w, dtype)
+    return mod, data, weight
+
+
+def tune_and_check(mod, data, weight):
```
Comment (comaniac) [inline tests/python/relay/test_auto_scheduler_layout_rewrite.py:73]: If we only want to test layout rewrite, we should be able to hard code a log without runnning a tuning on the fly?

File: python/tvm/auto_scheduler/compute_dag.py [hunk 6]
```diff
@@ -162,6 +162,23 @@ def infer_bound_from_state(self, state):
                 updated_state.stage_id_map[k] = v
         return updated_state
 
+    def rewrite_layout_from_state(self, state):
+        """
+        Rewrite the layout according to the transform steps in the history of a state
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/compute_dag.py:167]:
```suggestion
        Rewrite the layout of the DAG according to the transform steps
        in the history of a state.
```

File: tests/python/relay/test_auto_scheduler_layout_rewrite.py (hunk 5 above)
Comment (merrymercy) [reply tests/python/relay/test_auto_scheduler_layout_rewrite.py:73]: Regenerate them makes the test more comprehensive. It also makes the future updates of the log format easier.

File: python/tvm/auto_scheduler/relay_integration.py [hunk 7]
```diff
@@ -247,9 +271,74 @@ def auto_schedule_topi(outs, has_complex_op):
...
         schedule = te.create_schedule([x.op for x in outs])
     elif env.tracing_mode == TracingMode.PREPARE_LAYOUT_REWRITE:
-        # todo(merrymercy, minminsun): port layout rewrite
-        raise NotImplementedError
+        # in prepare_layout_rewrite mode
+        if enable_layout_rewrite and has_layout_free:
+            dispatch_ctx = DispatchContext.current
+            state = dispatch_ctx.query(target, key, has_complex_op, dag)
+            if state is None:
+                return te.create_schedule([x.op for x in outs])
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/relay_integration.py:279]:
```suggestion
                return None
```

Comment (merrymercy) [inline python/tvm/auto_scheduler/relay_integration.py:279]:
```suggestion
                return None
```

[the remark this answers is not in the document]
Comment (merrymercy) [reply python/tvm/auto_scheduler/relay_integration.py:279]: This does not matter. The return value is never used.

File: python/tvm/auto_scheduler/relay_integration.py [hunk 8]
```diff
@@ -247,9 +271,74 @@ def auto_schedule_topi(outs, has_complex_op):
...
+    tensor.__call__ = tensor_no_check_call.__get__(tensor, Tensor)
+
+
+def rewrite_compute_body(compute_tensor, new_layout):
+    """Rewrite the body of a ComputeOp according to a new layout of a placeholder"""
+    op = compute_tensor.op
+
+    # Get layout free placeholders
+    layout_free_placeholders = op.attrs["layout_free_placeholders"]
+    assert len(layout_free_placeholders) == 1
```
Comment (merrymercy) [inline python/tvm/auto_scheduler/relay_integration.py:333]:
```suggestion
    assert len(layout_free_placeholders) == 1, "Only support one layout free placeholders"
```

**Review by comaniac — APPROVED** [review]

File: include/tvm/topi/transform.h [hunk 9]
```diff
@@ -1400,6 +1400,74 @@ inline Tensor layout_transform(const Tensor& src, const std::string& src_layout,
       name, tag);
 }
 
+/*! \brief utility function for auto_scheduler_layout_transform */
```
Comment (merrymercy) [inline include/tvm/topi/transform.h:1403]:
```suggestion
/*! \brief Utility function for auto_scheduler_layout_transform */
```

File: include/tvm/relay/transform.h [hunk 10]
```diff
@@ -315,6 +323,12 @@ TVM_DLL Pass CanonicalizeOps();
  */
 TVM_DLL Pass AlterOpLayout();
 
+/*!
+ * \brief Do layout rewrite according to the tile structure created by auto-scheduler.
+ * \return The pass.
```
Comment (merrymercy) [inline include/tvm/relay/transform.h:328]:
```suggestion
 * \return The pass
```

**Pull request merged**